File: Mage/LocationDataStore.m
//
//  PeopleDataStore.m
//  MAGE
//
//

#import "LocationDataStore.h"
#import "PersonTableViewCell.h"
#import "Location.h"
#import "Locations.h"
#import "NSDate+DateTools.h"

@interface LocationDataStore ()
    @property (weak, nonatomic) IBOutlet UIViewController *viewController;
    @property (nonatomic) NSDateFormatter *dateFormatter;
    @property (nonatomic) NSDateFormatter *dateFormatterToDate;
@end

@implementation LocationDataStore

- (NSDateFormatter *) dateFormatter {
    if (_dateFormatter == nil) {
        _dateFormatter = [[NSDateFormatter alloc] init];
        _dateFormatter.dateStyle = kCFDateFormatterLongStyle;
        _dateFormatter.timeZone = [NSTimeZone timeZoneWithAbbreviation:@"UTC"];
    }
    
    return _dateFormatter;
    
}

- (NSDateFormatter *) dateFormatterToDate {
    if (_dateFormatterToDate == nil) {
        _dateFormatterToDate = [[NSDateFormatter alloc] init];
        _dateFormatterToDate.dateFormat = @"yyyy-MM-dd";
        _dateFormatterToDate.timeZone = [NSTimeZone timeZoneWithAbbreviation:@"UTC"];
    }
    
    return _dateFormatterToDate;
    
}

- (void) startFetchController {
    self.locations = [Locations locationsForAllUsers];
    self.locations.delegate = self;
    
    NSError *error;
    if (![self.locations.fetchedResultsController performFetch:&error]) {
        NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
    }
    
    [self.tableView reloadData];
}

- (void) updatePredicates {
    [self.locations.fetchedResultsController.fetchRequest setPredicate:[NSCompoundPredicate andPredicateWithSubpredicates:[Locations getPredicatesForLocations]]];
    NSError *error;
    if (![self.locations.fetchedResultsController performFetch:&error]) {
        NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
    }
    
    [self.tableView reloadData];
}

- (Location *) locationAtIndexPath: (NSIndexPath *)indexPath {
    return [self.locations.fetchedResultsController objectAtIndexPath:indexPath];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    id  sectionInfo = [[self.locations.fetchedResultsController sections] objectAtIndex:section];
    return [sectionInfo numberOfObjects];
}

- (void) configureCell:(UITableViewCell *) cell atIndexPath:(NSIndexPath *)indexPath {
	PersonTableViewCell *personCell = (PersonTableViewCell *) cell;
	
	Location *location = [self.locations.fetchedResultsController objectAtIndexPath:indexPath];
    [personCell populateCellWithUser:location.user];
    personCell.userActionsDelegate = self;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
	
    PersonTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"personCell"];
	
	[self configureCell:cell atIndexPath:indexPath];
	
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    PersonTableViewCell *cell = (PersonTableViewCell *)[tableView dequeueReusableCellWithIdentifier:@"personCell"];
    return cell.frame.size.height;
}

- (NSInteger) numberOfSectionsInTableView:(UITableView *)tableView {
    return [[self.locations.fetchedResultsController sections] count];
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    id <NSFetchedResultsSectionInfo> theSection = [[self.locations.fetchedResultsController sections] objectAtIndex:section];
    NSDate *date = [self.dateFormatterToDate dateFromString:[theSection name]];
    return [self.dateFormatter stringFromDate:date];
}

- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller {
    // The fetch controller is about to start sending change notifications, so prepare the table view for updates.
    [self.tableView beginUpdates];
}

- (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id) anObject atIndexPath:(NSIndexPath *) indexPath forChangeType:(NSFetchedResultsChangeType)type newIndexPath:(NSIndexPath *) newIndexPath {
	
    UITableView *tableView = self.tableView;
	
    switch(type) {
			
        case NSFetchedResultsChangeInsert:
            [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath] withRowAnimation:UITableViewRowAnimationNone];
            break;
			
        case NSFetchedResultsChangeDelete:
            [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationNone];
            break;
			
        case NSFetchedResultsChangeUpdate:
            [tableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
            break;
			
        case NSFetchedResultsChangeMove:
			[tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationNone];
			[tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath] withRowAnimation:UITableViewRowAnimationNone];
			break;
    }
}

- (void) controller:(NSFetchedResultsController *)controller didChangeSection:(id) sectionInfo atIndex:(NSUInteger) sectionIndex forChangeType:(NSFetchedResultsChangeType) type {
	
    switch(type) {
        case NSFetchedResultsChangeInsert:
            [self.tableView insertSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:UITableViewRowAnimationFade];
            break;
			
        case NSFetchedResultsChangeDelete:
            [self.tableView deleteSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:UITableViewRowAnimationFade];
            break;
        default:
            break;
    }
}

- (void) controllerDidChangeContent:(NSFetchedResultsController *)controller {
    // The fetch controller has sent all current change notifications, so tell the table view to process all updates.
    [self.tableView endUpdates];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    Location *location = [self.locations.fetchedResultsController objectAtIndexPath:indexPath];
    if (self.personSelectionDelegate) {
        [self.personSelectionDelegate userDetailSelected:location.user];
    }
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

- (void) userMapTapped:(PersonTableViewCell *) tableViewCell {
    NSIndexPath *indexPath = [self.tableView indexPathForCell:tableViewCell];
    Location *location = [self.locations.fetchedResultsController objectAtIndexPath:indexPath];
    if (self.personSelectionDelegate) {
        [self.personSelectionDelegate selectedUser:location.user];
    }
}

- (UIView *)tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger)section {
    UIView *view = [[UIView alloc] init];
    
    return view;
}

@end

File: Mage/LocationDataStore.swift
//
//  LocationDataStore.swift
//  MAGE
//
//  Created by Daniel Barela on 7/14/21.
//  Copyright © 2021 National Geospatial Intelligence Agency. All rights reserved.
//

import Foundation

class LocationDataStore: NSObject {
    
    var tableView: UITableView;
    var scheme: MDCContainerScheming?;
    var locations: Locations?;
    var actionsDelegate: UserActionsDelegate?;
    
    public init(tableView: UITableView, actionsDelegate: UserActionsDelegate?, scheme: MDCContainerScheming?) {
        self.scheme = scheme;
        self.tableView = tableView;
        self.actionsDelegate = actionsDelegate;
        super.init();
        self.tableView.dataSource = self;
        self.tableView.delegate = self;
    }
    
    func applyTheme(withContainerScheme containerScheme: MDCContainerScheming!) {
        self.scheme = containerScheme;
    }
    
    func startFetchController(locations: Locations? = nil) {
        if (locations == nil) {
            self.locations = Locations.forAllUsers();
        } else {
            self.locations = locations;
        }
        self.locations?.delegate = self;
        do {
            try self.locations?.fetchedResultsController.performFetch()
        } catch {
            print("Error fetching locations \(error) \(error.localizedDescription)")
        }
        self.tableView.reloadData();
    }
    
    func updatePredicates() {
        self.locations?.fetchedResultsController.fetchRequest.predicate = NSCompoundPredicate(andPredicateWithSubpredicates: Locations.getPredicatesForLocations() as! [NSPredicate]);
        do {
            try self.locations?.fetchedResultsController.performFetch()
        } catch {
            print("Error fetching users \(error) \(error.localizedDescription)")
        }
        self.tableView.reloadData();
    }
}

extension LocationDataStore: UITableViewDataSource {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let sectionInfo: NSFetchedResultsSectionInfo? = self.locations?.fetchedResultsController.sections?[section];
        return sectionInfo?.numberOfObjects ?? 0;
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return self.locations?.fetchedResultsController.sections?.count ?? 0;
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeue(cellClass: PersonTableViewCell.self, forIndexPath: indexPath);
        configure(cell: cell, at: indexPath);
        return cell;
    }
    
    func configure(cell: PersonTableViewCell, at indexPath: IndexPath) {
        if let location: Location = self.locations?.fetchedResultsController.object(at: indexPath) {
            cell.configure(location: location, actionsDelegate: actionsDelegate, scheme: scheme);
        }
    }
}

extension LocationDataStore: UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return CGFloat.leastNormalMagnitude;
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return CGFloat.leastNormalMagnitude;
    }
    
    func tableView(_ tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
        return UIView();
    }
    
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        return UIView();
    }
}

extension LocationDataStore: NSFetchedResultsControllerDelegate {
    
    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange anObject: Any, at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?) {
        switch type {
        case .insert:
            if let insertIndexPath = newIndexPath {
                self.tableView.insertRows(at: [insertIndexPath], with: .fade);
            }
            break;
        case .delete:
            if let deleteIndexPath = indexPath {
                self.tableView.deleteRows(at: [deleteIndexPath], with: .fade);
            }
            break;
        case .update:
            if let updateIndexPath = indexPath {
                self.tableView.reloadRows(at: [updateIndexPath], with: .none);
            }
            break;
        case .move:
            if let deleteIndexPath = indexPath {
                self.tableView.deleteRows(at: [deleteIndexPath], with: .fade);
            }
            if let insertIndexPath = newIndexPath {
                self.tableView.insertRows(at: [insertIndexPath], with: .fade);
            }
            break;
        default:
            break;
        }
    }
    
    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange sectionInfo: NSFetchedResultsSectionInfo, atSectionIndex sectionIndex: Int, for type: NSFetchedResultsChangeType) {
        switch type {
        case .insert:
            self.tableView.insertSections(IndexSet(integer: sectionIndex), with: .fade);
            break;
        case .delete:
            self.tableView.deleteSections(IndexSet(integer: sectionIndex), with: .fade);
            break;
        default:
            break;
        }
    }
    
    func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        self.tableView.endUpdates();
    }
    
    func controllerWillChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        self.tableView.beginUpdates();
    }
}
